benches/backtracking.rs

```diff
@@ -0,0 +1,52 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use std::collections::BTreeMap;
+
+use criterion::*;
+use pubgrub::OfflineDependencyProvider;
+use version_ranges::Ranges;
+
+type Deps = BTreeMap<u32, BTreeMap<u8, BTreeMap<u32, Vec<(u8, u8)>>>>;
+
+fn bench(b: &mut Bencher, data: &str) {
+    let dependencies = ron::de::from_str::<Deps>(data).unwrap();
+
+    let mut dependency_provider = OfflineDependencyProvider::<u32, Ranges<u8>>::new();
+
+    for (&p, vmap) in &dependencies {
+        for (&v, dmap) in vmap {
+            let deps_iter = dmap.iter().map(|(&d, intervals)| {
+                let mut r = Ranges::empty();
+                for &(start, end) in intervals {
+                    r = r.r#union(&Ranges::from_range_bounds(start..=end));
+                }
+                (d, r)
+            });
+            dependency_provider.add_dependencies(p, v, deps_iter);
+        }
+    }
+
+    if let Some((p, v)) = dependencies
+        .first_key_value()
+        .and_then(|(&p, vmap)| vmap.first_key_value().map(|(&v, _)| (p, v)))
```

Comment:
It would be really nice if this identified "the problematic crate from this file" more directly than "the one with the smallest index". I get that if this is generated from a arena based system package `0` version `0` has to be the root of the resolution. But, I definitely isolated real-world test cases where a non-root package was actually slower than the original problem.

Time resolve each package/version, and benchmark the slowest one would work. But timing is so unreliable without running things repeatedly. Count the number of `should cancel` calles might work, as a proxy for time. Depending on where were spending our time. If we exposed a way to count the number of backtracks, we could use that as a proxy. Even if it's not "always the correct answer" this it at least does what the name of this benchmark claims.

Reply:
> It would be really nice if this identified "the problematic crate from this file"

This bench comes from the resolution of `solana-archiver-lib v1.1.12` for cargo, and I already know this is the slowest crate for pubgrub (from your benchmark at https://github.com/Eh2406/pubgrub-crates-benchmark/issues/6#issuecomment-2402992814). So for me there is no need to try resolving for all its dependencies, which will only slows the bench.

Reply:
I have updated the bench files which are now synthetic. I didn't include the generation code in this PR because it needs code from #274. Here is the code for reference: [gen.rs.tar.gz](https://github.com/user-attachments/files/17689720/gen.rs.tar.gz).

Reply:
If it's synthetic, I think I'd rather have the generator than the generated file. It also needs good documentation about:
- Why it is pathological? Why does the structure force the resolver to spend it's time on what.
- What does it have to do with real cases? What real  situation induced this kind of pathological encoding.

The existing "large case"  file does not do well on either of these standards. It is synthetic.  It was pathological due to problems that have long since been fixed.  The only thing real world about it is that it is the kind of graph our prop-test  tends to generate.

The new sudoku benchmarks  are semisynthetic.  They are two real ways to encode a real problem, which makes them non-synthetic. On the other hand using a  pubgrub is a sudoku solver is pretty odd behavior by itself. The  fast encoding run so smoothly that it doesn't have a pathology to document. The slow encoding is slow because  every  version depends on a different package than the previous versions.  Which is not the data pattern pubgrub is designed for.